solar_consumer/save/save_data_platform.py

```diff
@@ -0,0 +1,191 @@
+
+        # 6. Get the most recent observations for that location and observer
+        recent_observations_request = dp.GetObservationsAsTimeseriesRequest(
+            location_uuid=location_uuid,
+            energy_source=dp.EnergySource.SOLAR,
+            observer_name=name,
+            time_window=dp.TimeWindow(
+                start_timestamp_utc=start_timestamp_utc,
+                end_timestamp_utc=end_timestamp_utc,
+            ),
+        )
+        recent_observations = await client.get_observations_as_timeseries(
+            recent_observations_request
+        )
+
+        # 7. Remove any data points from our data that are already in the database
+        logger.debug(
+            f"Found {len(recent_observations.values)} recent observations for location {location_uuid} and observer {name}."
+        )
+        data_in_database = {
+            obs.timestamp_utc: obs.value_fraction for obs in recent_observations.values
+        }
+        gsp_data = gsp_data[~gsp_data["target_datetime_utc"].isin(data_in_database.keys())]
+
+        if len(gsp_data) == 0:
+            logger.debug(
+                f"No new data to add for location {location_uuid} and observer {name}, for gsp id {gsp_id}."
+            )
+            continue
+
+        # 8. Update location capacity based on the max capacity in this data
+        new_max_capacity_watts = int(gsp_data["capacity_mwp"].max() * 1_000_000)
+        max_capacity_watts_current = location.effective_capacity_watts
+        if new_max_capacity_watts > max_capacity_watts_current:
+            logger.info(
+                f"Updating location {location_uuid} capacity from {max_capacity_watts_current}W to {new_max_capacity_watts}W."
+            )
+
+            update_location_request = dp.UpdateLocationCapacityRequest(
+                location_uuid=location_uuid,
+                energy_source=dp.EnergySource.SOLAR,
+                valid_from_utc=pd.Timestamp.utcnow().to_pydatetime(),
+                new_effective_capacity_watts=new_max_capacity_watts,
+            )
+            _ = await client.update_location_capacity(update_location_request)
+        else:
+            logger.debug(
+                f"Location {location_uuid} capacity of {max_capacity_watts_current}W is sufficient; no update needed."
+            )
+
+        # 9. Create new observations for the remaining data points
+        observation_values = []
+        for _, row in gsp_data.iterrows():
+            if row["capacity_mwp"] == 0:
+                continue  # Skip entries with zero capacity to avoid division by zero
+
+            value_fraction = row["solar_generation_kw"] / (row["capacity_mwp"] * 1000)
+            effective_capacity_watts = int(row["capacity_mwp"] * 1_000_000)
+            timestamp_utc = row["target_datetime_utc"].to_pydatetime()
+
+            oberservation_value = dp.CreateObservationsRequestValue(
+                timestamp_utc=timestamp_utc,
```

Comment:
Out of interest, what is the accuracy on these timestamps? Are they clipped to the nearest mnute/second? If not, it might be worth doing so, for ease of comparison with predicted values